### DISCUSSION TITLE
Why drop OpenGL?

### DISCUSSION THREAD

**Discussion by amerkoleci** [bjornbytes/lovr#664] [General]
Hi,
Excellent job on this library, is one of those projects that I check from time to time.z

I was wondering, why opengl/webgl was dropped? I saw there would be WebGPU backend but I suppose lot of implementations won't catch with release on all platforms. 

Cheers

**Comment by shakesoda** [comment]
off the top of my head:

- needing to account for pre-compute apis is crippling
- opengl is horrendously bad with threading
- opengl has terribly inconsistent performance and means to mitigate it are merely hopes and suggestions
- opengl has wildly inconsistent support levels (ranging between "stop using this for new code" and "stop using this 15 years ago for new code")
- all the hardware lovr is intended to support (i.e. vr-capable hardware) supports vulkan
- maintaining the opengl backend along with designing new ones is an enormous maintenance burden unless you make both of them terrible

losing the web target for now is a notable loss, but it'll come back.

**Comment by bjornbytes** [comment]
@shakesoda's diplomatic summary raises good points.  I will add that during development I initially tried to keep the OpenGL renderer intact, but it proved to be too difficult to maintain both, especially since the design/architecture was changing very rapidly as I was still learning Vulkan and how I wanted to structure things.

Now that the dust has (mostly) settled it would probably be possible to add an OpenGL backend.  However I don't see a reason to do so since it would increase maintenance burden, the time could be spent on more important things (like WebGPU), and writing OpenGL is not very fun for me anymore after learning Vulkan (stockholm syndrome?).

With that said, there have been at least 2 people since v0.16.0 who tried to use LÖVR but couldn't, because their integrated GPUs don't support Vulkan.  This kind of sucks, but on the other hand this hardware doesn't support VR anyway.
